src/packet/signature.js

```diff
@@ -654,41 +651,58 @@ class SignaturePacket {
    * @param {String|Object} data - Data which on the signature applies
+   * @param {Date} [date] - Use the given date instead of the current time to check for signature validity and expiration
    * @param {Boolean} [detached] - Whether to verify a detached signature
    * @param {Object} [config] - Full configuration, defaults to openpgp.config
    * @throws {Error} if signature validation failed
    * @async
    */
-  async verify(key, signatureType, data, detached = false, config = defaultConfig) {
+  async verify(key, signatureType, data, date = new Date(), detached = false, config = defaultConfig) {
     const publicKeyAlgorithm = enums.write(enums.publicKey, this.publicKeyAlgorithm);
     const hashAlgorithm = enums.write(enums.hash, this.hashAlgorithm);
-
+    if (!this.issuerKeyID.equals(key.getKeyID())) {
+      throw new Error('Signature was not issued by the given public key');
+    }
     if (publicKeyAlgorithm !== enums.write(enums.publicKey, key.algorithm)) {
       throw new Error('Public key algorithm used to sign signature does not match issuer key algorithm.');
     }
 
-    let toHash;
-    let hash;
-    if (this.hashed) {
-      hash = await this.hashed;
-    } else {
-      toHash = this.toHash(signatureType, data, detached);
-      hash = await this.hash(signatureType, data, toHash);
-    }
-    hash = await stream.readToEnd(hash);
-    if (this.signedHashValue[0] !== hash[0] ||
-        this.signedHashValue[1] !== hash[1]) {
-      throw new Error('Message digest did not match');
-    }
+    const isMessageSignature = signatureType === enums.signature.binary || signatureType === enums.signature.text;
+    // Cryptographic validity is cached after one successful verification.
+    // However, for message signatures, we always re-verify, since the passed `data` can change
+    const skipVerify = this.verified && !isMessageSignature;
+    if (!skipVerify) {
+      let toHash;
+      let hash;
+      if (this.hashed) {
+        hash = await this.hashed;
+      } else {
+        toHash = this.toHash(signatureType, data, detached);
+        hash = await this.hash(signatureType, data, toHash);
+      }
+      hash = await stream.readToEnd(hash);
+      if (this.signedHashValue[0] !== hash[0] ||
+          this.signedHashValue[1] !== hash[1]) {
+        throw new Error('Signed digest did not match');
+      }
 
-    this.params = await this.params;
+      this.params = await this.params;
 
-    const verified = await crypto.signature.verify(
-      publicKeyAlgorithm, hashAlgorithm, this.params, key.publicParams,
-      toHash, hash
-    );
-    if (!verified) {
-      throw new Error('Signature verification failed');
+      const verified = await crypto.signature.verify(
+        publicKeyAlgorithm, hashAlgorithm, this.params, key.publicParams,
+        toHash, hash
+      );
+      if (!verified) {
+        throw new Error('Signature verification failed');
+      }
```

Comment:
I'd rather not mark a signature as verified when e.g. we consider the hash algo to be unsafe. I actually think it might be best to reset the flag once we detect that the signature is no longer valid, because we do return the signature packets as part of `signatures` in the top-level functions, and people might end up accessing `signatures[0].signature.verified`